src/qibo/models/dbi/double_bracket.py

```diff
@@ -156,6 +203,98 @@ def hyperopt_step(
         )
         return best["step"]
 
+    def polynomial_step(
+        self,
+        n: int = 4,
+        n_max: int = 5,
+        d: np.array = None,
+        backup_scheduling: DoubleBracketScheduling = None,
+    ):
+        r"""
+        Optimizes iteration step by solving the n_th order polynomial expansion of the loss function.
+        e.g. $n=2$: $2\Trace(\sigma(\Gamma_1 + s\Gamma_2 + s^2/2\Gamma_3)\sigma(\Gamma_0 + s\Gamma_1 + s^2/2\Gamma_2))
+        Args:
+            n (int, optional): the order to which the loss function is expanded. Defaults to 4.
+            n_max (int, optional): maximum order allowed for recurring calls of `polynomial_step`. Defaults to 5.
+            d (np.array, optional): diagonal operator, default as $\delta(H)$.
+            backup_scheduling (`DoubleBracketScheduling`): the scheduling method to use in case no real positive roots are found.
+        """
+
+        if d is None:
+            d = self.diagonal_h_matrix
+
+        if backup_scheduling is None:
+            backup_scheduling = DoubleBracketScheduling.grid_search
+
+        def sigma(h: np.array):
+            return h - self.backend.cast(np.diag(np.diag(self.backend.to_numpy(h))))
+
+        def Gamma(k: int):
+            r"""Computes the k_th Gamma function i.e $\Gamma_k=[W,...,[W,[W,H]]...]$, where we take k nested commutators with $W = [D, H]$"""
+            if k == 0:
+                return self.h.matrix
+            else:
+                W = self.commutator(d, sigma(self.h.matrix))
+                result = self.h.matrix
+                for _ in range(k):
+                    result = self.commutator(W, result)
+            return result
+
+        # list starting from s^n highest order to s^0
+        sigma_gamma_list = np.array([sigma(Gamma(k)) for k in range(n + 2)])
+        exp_list = np.array([1 / math.factorial(k) for k in range(n + 1)])
+        # coefficients for rotation with [W,H] and H
+        c1 = [
+            exp_coef * delta_gamma
+            for exp_coef, delta_gamma in zip(exp_list, sigma_gamma_list[1:])
+        ]
```

Comment:
If `exp_list` and sigma_gamma_list are numpy arrays, you can just multiply them
```suggestion
        c1 = exp_list * sigma_gamma_list[1:]
```

Reply:
Modified: `c1 = exp_list.reshape(-1,1,1) * sigma_gamma_list[1:]` to match the shape